src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -81,37 +81,447 @@ constexpr const ad_utility::Last<Current, Others...>* unwrapVariant(
 
 // ___________________________________________________________________________
 using PathTuples = ad_utility::sparql_types::PathTuples;
+using VarOrPath = ad_utility::sparql_types::VarOrPath;
+using ObjectList = ad_utility::sparql_types::ObjectList;
+using TripleWithPropertyPath = ad_utility::sparql_types::TripleWithPropertyPath;
+
+vector<TripleWithPropertyPath> SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::TriplesSameSubjectPathContext* ctx) {
+  vector<TripleWithPropertyPath> triples;
+  if (ctx->varOrTerm()) {
+    auto subject = visitTypesafe(ctx->varOrTerm());
+    auto tuples = visitTypesafe(ctx->propertyListPathNotEmpty());
+    for (auto& tuple : tuples) {
+      triples.push_back(
+          TripleWithPropertyPath{subject, tuple.first, tuple.second});
+    }
+  } else if (ctx->triplesNodePath()) {
+    throwCollectionsAndBlankNodePathsNotSupported(ctx->triplesNodePath());
+  }
+  return triples;
+}
+
+// ___________________________________________________________________________
+std::optional<PathTuples> SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::PropertyListPathContext* ctx) {
+  if (ctx->propertyListPathNotEmpty()) {
+    return visitTypesafe(ctx->propertyListPathNotEmpty());
+  } else {
+    return std::nullopt;
+  }
+}
+
+// ___________________________________________________________________________
+PropertyPath SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::VerbPathContext* ctx) {
+  PropertyPath p = visitTypesafe(ctx->path());
+  p.computeCanBeNull();
+  return p;
+}
+
+// ___________________________________________________________________________
+Variable SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::VerbSimpleContext* ctx) {
+  return visitTypesafe(ctx->var());
+}
+
+// ___________________________________________________________________________
+ObjectList SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::ObjectListPathContext* ctx) {
+  // The second parameter is empty because collections and blank not paths,
+  // which might add additional triples, are currently not supported.
+  return ObjectList{visitVector<VarOrTerm>(ctx->objectPath()), {}};
+}
+
+// ___________________________________________________________________________
+PropertyPath SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::PathContext* ctx) {
+  return visitTypesafe(ctx->pathAlternative());
+}
+
+// ___________________________________________________________________________
+PropertyPath SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::PathAlternativeContext* ctx) {
+  auto paths = visitVector<PropertyPath>(ctx->pathSequence());
+
+  if (paths.size() == 1) {
+    return paths[0];
+  } else {
+    return PropertyPath::makeAlternative(std::move(paths));
+  }
```

Comment:
Maybe move this optimization into the `makeAlternative...` function 
(similar for the `makeSequence` below, to make the visitor as simple as trivial as possible.